ARROW-18238: [Docs][Python] Improve docs for S3FileSystem (#14599)
apache/arrow: docs/source/python/filesystems.rst

   >>> f.readall()
   b'some data'

>    # Note, to resolve the S3 bucket, one can use `resolve_s3_region('my-test-bucket')`

Reviewer: I would move this to an actual text paragraph instead of "hiding" it somewhat in a comment in a longer example. And maybe first stating that the S3FileSystem needs to be configured with the correct region for the bucket you want to read. And then if you want to do that automatically instead of passing the region manually (as in the example above), explain and show there are two ways you can do that.

Author: Indeed, I think that's better. :+1:

   >>> s3 = fs.S3FileSystem(region=resolve_s3_region('my-test-bucket'))

   # Or alternatively...

>    >>> s3 = fs.S3FileSystem.from_uri('s3://[access_key:secret_key@]bucket/path[?region=]')

Reviewer: ```suggestion
   >>> s3, path = fs.S3FileSystem.from_uri('s3://[access_key:secret_key@]bucket/path[?region=]')
```

`from_uri` returns two elements. 

Also, maybe leave out the `[?region]` part, since the goal here is to show that `from_uri` will infer the region from the bucket name (without the need to manually specify it)

   >>> f.readall()
   b'some data'

Note that it is important to configure :class:`S3FileSystem` with the correct
region for the bucket being used. If `region` is not set, the AWS SDK will
choose a value, defaulting to 'us-east-1' if the SDK version is <1.8.

> Otherwise it'll try to use a variety of heuristics to resolve the region.

Reviewer: I would maybe add the "(environment variables, configuration profile, EC2 metadata server)" (copied from the C++ source), to make it clear that the "heuristics" don't include "infer it from the region"

```suggestion
Otherwise it will try to use a variety of heuristics to resolve the region.
```

> It is also possible to resolve the region for :class:`S3FileSystem` by using

Reviewer: ```suggestion
It is also possible to resolve the region from the bucker name for :class:`S3FileSystem` by using
```
